runtime/lua/vim/lsp/document_color.lua

```diff
@@ -0,0 +1,296 @@
+--- @field bufnr? integer
+
+--- Enables document highlighting from the given language client in the given buffer.
+---
+--- To "toggle", pass the inverse of `is_enabled()`:
+---
+--- ```lua
+--- lsp.document_color.enable(not vim.lsp.document_color.is_enabled())
+--- ```
+---
+--- @param enable? boolean True to enable, false to disable
+--- @param filter? lsp.document_color.enable.Filter
+function M.enable(enable, filter)
+  vim.validate('enable', enable, 'boolean', true)
+  vim.validate('filter', filter, 'table', true)
+
+  enable = enable == nil or enable
+  filter = filter or {}
+
+  if filter.bufnr == nil then
+    global_state.enabled = enable
+    for _, bufnr in ipairs(api.nvim_list_bufs()) do
+      if api.nvim_buf_is_loaded(bufnr) then
+        if enable then
+          buf_enable(bufnr)
+        else
+          buf_disable(bufnr)
+        end
+      else
+        bufstates[bufnr] = nil
+      end
+    end
+  else
+    local bufnr = vim._resolve_bufnr(filter.bufnr)
+    if enable then
+      buf_enable(bufnr)
+    else
+      buf_disable(bufnr)
+    end
+  end
+end
+
+api.nvim_create_autocmd('LspNotify', {
+  group = document_color_augroup,
+  callback = function(args)
+    local bufnr = args.buf
+    local method = args.data.method --- @type string
+
+    if
+      (method == ms.textDocument_didChange or method == ms.textDocument_didOpen)
+      and bufstates[bufnr].enabled
+    then
+      buf_refresh(bufnr, { client_id = args.data.client_id })
+    end
+  end,
+})
+
+api.nvim_create_autocmd('LspAttach', {
+  group = document_color_augroup,
+  callback = function(args)
+    api.nvim_buf_attach(args.buf, false, {
+      on_reload = function(_, bufnr)
+        buf_clear(bufnr)
+        if bufstates[bufnr].enabled then
+          buf_refresh(bufnr)
+        end
+      end,
+      on_detach = function(_, bufnr)
+        buf_disable(bufnr)
+      end,
+    })
+  end,
+})
+
+api.nvim_create_autocmd('LspDetach', {
+  group = document_color_augroup,
+  callback = function(args)
+    local clients = lsp.get_clients({ bufnr = args.buf, method = ms.textDocument_documentColor })
+
+    if not vim.iter(clients):any(function(c)
+      return c.id ~= args.data.client_id
+    end) then
+      -- There are no clients left in the buffer that support document color, so turn it off.
+      buf_disable(args.buf)
+    end
+  end,
+})
+
+api.nvim_set_decoration_provider(document_color_ns, {
+  on_win = function(_, _, bufnr)
+    local bufstate = rawget(bufstates, bufnr) --- @type vim.lsp.document_color.BufState
+
+    if
+      not bufstate
+      or not bufstate.hl_infos
+      or bufstate.buf_version ~= util.buf_versions[bufnr]
+      or bufstate.applied_version == bufstate.buf_version
+    then
+      return
+    end
+
+    for _, client_hls in pairs(bufstate.hl_infos) do
+      for _, hl in ipairs(client_hls) do
+        api.nvim_buf_set_extmark(bufnr, document_color_ns, hl.range[1], hl.range[2], {
```

Comment:
If you're using non-ephemeral marks in a decor provider, then you need to clear the namespace at the beginning of `on_win`. Alternatively you can make these marks ephemeral.